src/config.rs

```diff
@@ -0,0 +1,242 @@
+/* SPDX-License-Identifier: MIT */
+
+use anyhow::{anyhow, Context, Result};
+use crate::generator::run_generator;
+use crate::setup::run_device_setup;
+use ini::ini::{Ini, Properties as IniProperties, SectionIntoIter};
+use std::borrow::Cow;
+use std::cmp;
+use std::env;
+use std::fmt;
+use std::fs;
+use std::io::{prelude::*, BufReader};
+use std::iter::FromIterator;
+use std::path::{Path, PathBuf};
+
+
+pub struct Device {
+    pub name: String,
+    pub host_memory_limit_mb: Option<u64>,
+    pub zram_fraction: f64,
+    pub max_zram_size_mb: Option<u64>,
+    pub compression_algorithm: Option<String>,
+    pub disksize: u64,
+}
+
+impl Device {
+    fn new(name: String) -> Device {
+        Device {
+            name,
+            host_memory_limit_mb: Some(2 * 1024),
+            zram_fraction: 0.25,
+            max_zram_size_mb: Some(4 * 1024),
+            compression_algorithm: None,
+            disksize: 0,
+        }
+    }
+
+    fn write_optional_mb(f: &mut fmt::Formatter<'_>, val: Option<u64>) -> fmt::Result {
+        match val {
+            Some(val) => {
+                write!(f, "{}", val)?;
+                f.write_str("MB")?;
+            }
+            None => f.write_str("<none>")?,
+        }
+        Ok(())
+    }
+}
+
+impl fmt::Display for Device {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}: host-memory-limit=", self.name)?;
+        Device::write_optional_mb(f, self.host_memory_limit_mb)?;
+        write!(f, " zram-fraction={} max-zram-size=", self.zram_fraction)?;
+        Device::write_optional_mb(f, self.max_zram_size_mb)?;
+        f.write_str(" compression-algorithm=")?;
+        match self.compression_algorithm.as_ref() {
+            Some(alg) => f.write_str(alg)?,
+            None => f.write_str("<default>")?,
+        }
+        Ok(())
+    }
+}
+
+
+pub struct Config {
+    pub root: Cow<'static, str>,
+    pub module: ModuleConfig,
+}
+
+pub enum ModuleConfig {
+    Generator {
+        devices: Vec<Device>,
+        output_directory: PathBuf,
+    },
+    DeviceSetup {
+        device: Option<Device>,
+        name: String,
+    },
+}
+
+
+impl Config {
+    pub fn parse() -> Result<Config> {
+        let root: Cow<'static, str> = match env::var("ZRAM_GENERATOR_ROOT") {
+            Ok(val) => val.into(),
+            Err(env::VarError::NotPresent) => "/".into(),
+            Err(e) => return Err(e.into()),
+        };
+        println!("Using {:?} as a root directory", root);
```

Comment:
This is copied from `main.rs` in the split, so commenting here: I disagree with this stylistically – printing this on each run is hardly useful, it's better suited in the `Ok()` branch.